### ISSUE TITLE
Error registering fd event for the new client

### ISSUE THREAD

**Issue by piscopoc** [JohnSully/KeyDB#82]
Noticing the following errors in KeyDB log files after upgrading to version 5.0.2:
`
Error registering fd event for the new client: No such file or directory (fd=157)
epoll_ctl failed: No such file or directory
`

Seems to be related to issue #76. Do you have any guidance on this issue?

**Comment by JohnSully** [comment]
While both are fd related - this is part of the traditional client connection process. I don’t think its related to the PUB/SUB issues in the bug you linked because it occurs far too early for that code to run. 

Can you provide some background on the workload around when you see this error?  It will give some places to start looking while I investigate.

**Comment by piscopoc** [comment]
The error seems to popup fairly often in the logs, probably once every 30 seconds. We use Redis lists pretty heavily in production, it's not uncommon for a single list to see about 5 million items in a day.

Below is the replication information from one master to another:
`MASTER <-> REPLICA sync: receiving 5250553532 bytes from master
MASTER <-> REPLICA sync: Keeping old data
MASTER <-> REPLICA sync: Loading DB in memory
Loading RDB produced by version 5.0.2
RDB age 114 seconds
RDB memory usage when created 9210.11 Mb`

**Comment by piscopoc** [comment]
Seeing more of the same error today but this time with fd=274, not sure if that makes a difference or not.

```
epoll_ctl failed: No such file or directory
Error registering fd event for the new client: No such file or directory (fd=274)
```

We have haproxy in-front of our instances and it shows about 200 connections open currently. If you need any other information surrounding our configuration please let me know.

**Comment by JohnSully** [comment]
Could you temporarily enable debug level logging and post the logs around the error message?

You can do so with the command "config set loglevel debug".  The default level is "notice" for when you set it back.  This may temporarily impact performance but it will really help narrow down what is going on.  

There should not be an opportunity for the fd to be closed between when the socket is open and the client is created, the extra logs should help me trace execution and narrow it down more.

Also, are you using unix sockets at all or just TCP?

**Comment by JohnSully** [comment]
Also are you using any modules?

I'm getting a little closer to understanding the issue but I still need to narrow it down.  It looks like its not related to the fd closing at all, but the epoll state getting out of sync.

**Comment by piscopoc** [comment]
> Also, are you using unix sockets at all or just TCP?
Just TCP.

> Also are you using any modules?
No we aren't using any modules.

> Are you using clusters?
Yes we are using active-active replication between two instances.

I rebooted instance 1 and left instance 2 running and turned on debug logging. Not much to report as the two synced outside of a lot of "Connection reset by peer" messages on instance 1.

instance 1 logs:
Accepted 10.30.8.216:34738
Accepted 10.30.8.216:34740
Accepted 10.30.8.216:34742
Reading from client: Connection reset by peer
Reading from client: Connection reset by peer
Reading from client: Connection reset by peer
Reading from client: Connection reset by peer

instance 2 logs:
DB 0: 99588 keys (57166 volatile) in 131072 slots HT.
Accepted 10.30.8.216:46626
Accepted 10.30.8.216:46646
Accepted 10.30.8.216:46664
Accepted 10.30.8.216:46682
DB 0: 99569 keys (57147 volatile) in 131072 slots HT.
34 clients connected (1 replicas), 11153330720 bytes in use

I did see this in the logs though:
Client id=250648880 addr=10.30.29.189:39147 fd=89 name= age=103 idle=103 flags=S db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=425 oll=30286 omem=268440448 events=r cmd=psync scheduled to be closed ASAP for overcoming of output buffer limits.

**Comment by JohnSully** [comment]
The line:

> Client id=250648880 addr=10.30.29.189:39147 fd=89 name= age=103 idle=103 flags=S db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=425 oll=30286 omem=268440448 events=r cmd=psync scheduled to be closed ASAP for overcoming of output buffer limits.

Means that the replication backlog was too large to send to the client.  Its normal for it to disconnect and send the RDB when this happens.

If you are able to keep running with the debug log level I would need the 10 or so lines before the error message occurs:

> Error registering fd event for the new client: No such file or directory (fd=157) epoll_ctl failed: No such file or directory

In the meantime I will keep investigating.  I think I'm close to resolving the problem.

**Comment by JohnSully** [comment]
Are you using the append only file?  I might have found a potential issue in the AOF code path.

(Sorry for asking so many questions, there's just a lot of features to go through)

**Comment by piscopoc** [comment]
> Are you using the append only file? And if so do you ever manually call the BGREWRITEAOF command? I might have found a potential issue in the AOF code path.

Yes we are using the append only file across both instances. Yes the BGREWRITEAOF command is ran occasionally in that environment (not on a consistent basis, maybe once every other day).

**Comment by JohnSully** [comment]
Great! That means I have a fix that should resolve this for you.  Change: 55324aa56

Its currently in the unstable branch (unstable docker regenerates nightly), and will be included in the next release.

**Comment by JohnSully** [comment]
This bug is fixed and will be released in v5.1.  I'm aiming for Monday.
